realm/realm-library/src/main/java/io/realm/internal/android/JsonUtils.java

```diff
@@ -37,13 +41,26 @@
      */
     public static Date stringToDate(String date) {
         if (date == null || date.length() == 0) return null;
+
+        // Check for JSON date
         Matcher matcher = jsonDate.matcher(date);
         if (matcher.find()) {
             String dateMatch = matcher.group(1);
             return new Date(Long.parseLong(dateMatch));
-        } else {
+        }
+
+        // Check for millisecond based date
+        if(numericOnly.matcher(date).matches()) {
             return new Date(Long.parseLong(date));
         }
+
+        // Try for ISO8601 date
+        try {
+            parsePosition.setIndex(0); // reset the position each time
+            return ISO8601Utils.parse(date, parsePosition);
+        } catch (ParseException ignored) {
+            return null;
```

Comment:
This looks like a change from the current behaviour where we crash if the date isn't in a format we support. The old behaviour seems more safe IMO. This implementation could cause objects to be imported successfully but with wrong dates. In the old implementation we end up throwing a ParseNumberException, so just throwing the parse exception here would probably be fine.